### PR TITLE
Change AssignmentOperator to use IsRef instead of RefKind

### PR DESCRIPTION
**Pull request by agocke**
Assignments can be ref or regular, but there's no such thing as a
"ref-readonly" assignment or "out" assignment, only left-hand sides that
can be ref. Changing to IsRef prevents bugs around looking for the
RefKind of the assignment instead of looking for the RefKind of the LHS.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs [hunk 1]
```diff
@@ -102,7 +102,7 @@ internal static class MethodBodySynthesizer
                         objectType)
                     { WasCompilerGenerated = true },
                     thisReference,
-                    RefKind.None,
+                    false,
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs]: >false [](start = 20, length = 5)

A named argument would be nice.

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs [hunk 2]
```diff
@@ -5,11 +5,36 @@
...
+                case BoundKind.Local:
+                    return ((BoundLocal)node).LocalSymbol.RefKind;
+
+                case BoundKind.Parameter:
+                    return ((BoundParameter)node).ParameterSymbol.RefKind;
+
+                case BoundKind.Call:
+                    return ((BoundCall)node).Method.RefKind;
+
+                case BoundKind.PropertyAccess:
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs]: Do we need to handle ternary expression here as well?

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 3]
```diff
@@ -379,7 +379,7 @@
     through addr. We therefore need to disambiguate what kind of assignment
     we are doing based on something other than the refness of the left hand side. 
     -->
-    <Field Name="RefKind" Type="RefKind" Null="NotApplicable"/>
+    <Field Name="IsRef" Type="bool" Null="NotApplicable"/>
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: ConditionalOperator uses "IsByRef".   Would it make sense to use the same in both cases? 
It does not matter which one, but "IsRef" sounds a bit better to me.

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs (hunk 2 above)
Comment (VSadov) [reply src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs]: Oh, I see. The whole point here is that we don't.

---
In reply to: [153597330](https://github.com/dotnet/roslyn/pull/23415#discussion_r153597330) [](ancestors = 153597330)

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs [hunk 4]
```diff
@@ -5,11 +5,36 @@
...
 using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
     internal static partial class BoundExpressionExtensions
     {
+        /// <summary>
+        /// Only legal for expressions that can produce l-values.
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs]: description is too vague. ref ternaries, array accesses, field accesses also can produce lvalues. 

I understand what the helper is for, but it is not what the description says.

Comment (VSadov) [reply src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs]: It seems the helper is legal only for expressions that represent _symbols_ which can have RefKind.

---
In reply to: [153599228](https://github.com/dotnet/roslyn/pull/23415#discussion_r153599228) [](ancestors = 153599228)

File: src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.cs]: Unfortunately this is one of those rare cases where adding named arguments actually makes overload resolution ambiguous.

**Comment by agocke** [comment]
@jaredpar For ask-mode review.

**Comment by agocke** [comment]
@MeiChin-Tsai for ask mode approval

**Comment by MeiChin-Tsai** [comment]
approved

**Pull request merged**